salt/grains/core.py

```diff
@@ -2320,9 +2320,14 @@ def _legacy_linux_distribution_data(grains, os_release, lsb_has_error):
     log.trace(
         "Getting OS name, release, and codename from distro id, version, codename"
     )
-    (osname, osrelease, oscodename) = (
-        x.strip('"').strip("'") for x in _linux_distribution()
-    )
+    try:
+        (osname, osrelease, oscodename) = (
+            x.strip('"').strip("'") for x in _linux_distribution()
+        )
+    except subprocess.CalledProcessError:
+        # When running under the salt user, a call to `lsb_release` might raise
+        # permissin errors.
+        return grains
```

Comment:
Wondering about the information that won't be there if encounter this issue, e.g. oscodename, os, os_family especially since this is often used to decide RedHat family or Debian/Ubuntu family

Reply:
I understand that salt is not raising the exception.
I am stating that the code changes would be better located in salt/utils/platform.py and in the function linux_distribution, localizing the issue to that function, and fixing the issue for all users of that function.
In addition, perhaps there is a way to make a best guess as to name, version, codename or id in the exception path such that useful information is returned.
Either way, exception handling would be better situated in linux_distribution

Reply:
> I understand that salt is not raising the exception. I am stating that the code changes would be better located in salt/utils/platform.py and in the function linux_distribution, localizing the issue to that function, and fixing the issue for all users of that function. In addition, perhaps there is a way to make a best guess as to name, version, codename or id in the exception path such that useful information is returned. Either way, exception handling would be better situated in linux_distribution

And, if the exception is raised, what do we return?
Most importantly, what do we return, that the calling function can make use of.
I don't want to do any kind of parsing or guessing, that's why we are using `distro` in the first place...

Reply:
Always /etc/os-release :) , flippantly joking
At a minimum you need to handle use of linux_distribution in version.py too, but better to handle in one place, the piece of code with the problem.

Reply:
uname -v on Debian has Debian in the output
uname -v on Ubuntu has Ubuntu in the output
uname -v on Rhel 7 doesn't have anything stating RedHat in the output, but that could be used to assume Redhat os_family
And at least deciding Redhat or Debian/Ubuntu family if not already available would be a big win, since yum/dnf vs apt

Reply:
I don't want to re-invent the wheel, that's why we use distro.
I'll happily catch the permissions error and return an empty string(which is what ditro does quite often in its code)

```
root@ip-10-1-21-121 [ ~ ]# /tmp/testing/artifacts/salt/bin/python3
Python 3.10.12 (main, Aug 14 2023, 22:21:45) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import distro
>>> distro.codename()
'Photon'
>>> distro.name()
'VMware Photon OS'
>>> distro.version(best=True)
'4.0'
>>> distro.codename()
'Photon'
>>>
```
```
root@ip-10-1-21-121 [ ~ ]# sudo su - salt-local
salt-local [ ~ ]$ /tmp/testing/artifacts/salt/bin/python3
Python 3.10.12 (main, Aug 14 2023, 22:21:45) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import distro
>>> distro.name()
'VMware Photon OS'
>>> distro.version(best=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/tmp/testing/artifacts/salt/lib/python3.10/site-packages/distro.py", line 287, in version
    return _distro.version(pretty, best)
  File "/tmp/testing/artifacts/salt/lib/python3.10/site-packages/distro.py", line 741, in version
    self.lsb_release_attr('release'),
  File "/tmp/testing/artifacts/salt/lib/python3.10/site-packages/distro.py", line 903, in lsb_release_attr
    return self._lsb_release_info.get(attribute, '')
  File "/tmp/testing/artifacts/salt/lib/python3.10/site-packages/distro.py", line 556, in __get__
    ret = obj.__dict__[self._fname] = self._f(obj)
  File "/tmp/testing/artifacts/salt/lib/python3.10/site-packages/distro.py", line 1014, in _lsb_release_info
    stdout = subprocess.check_output(cmd, stderr=devnull)
  File "/tmp/testing/artifacts/salt/lib/python3.10/subprocess.py", line 421, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/tmp/testing/artifacts/salt/lib/python3.10/subprocess.py", line 526, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '('lsb_release', '-a')' returned non-zero exit status 126.
>>> distro.version()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/tmp/testing/artifacts/salt/lib/python3.10/site-packages/distro.py", line 287, in version
    return _distro.version(pretty, best)
  File "/tmp/testing/artifacts/salt/lib/python3.10/site-packages/distro.py", line 741, in version
    self.lsb_release_attr('release'),
  File "/tmp/testing/artifacts/salt/lib/python3.10/site-packages/distro.py", line 903, in lsb_release_attr
    return self._lsb_release_info.get(attribute, '')
  File "/tmp/testing/artifacts/salt/lib/python3.10/site-packages/distro.py", line 556, in __get__
    ret = obj.__dict__[self._fname] = self._f(obj)
  File "/tmp/testing/artifacts/salt/lib/python3.10/site-packages/distro.py", line 1014, in _lsb_release_info
    stdout = subprocess.check_output(cmd, stderr=devnull)
  File "/tmp/testing/artifacts/salt/lib/python3.10/subprocess.py", line 421, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/tmp/testing/artifacts/salt/lib/python3.10/subprocess.py", line 526, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '('lsb_release', '-a')' returned non-zero exit status 126.
>>> distro.codename()
Traceback (most recent call last):
  File "/tmp/testing/artifacts/salt/lib/python3.10/site-packages/distro.py", line 824, in codename
    return self._os_release_info['codename']
KeyError: 'codename'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/tmp/testing/artifacts/salt/lib/python3.10/site-packages/distro.py", line 384, in codename
    return _distro.codename()
  File "/tmp/testing/artifacts/salt/lib/python3.10/site-packages/distro.py", line 826, in codename
    return self.lsb_release_attr('codename') \
  File "/tmp/testing/artifacts/salt/lib/python3.10/site-packages/distro.py", line 903, in lsb_release_attr
    return self._lsb_release_info.get(attribute, '')
  File "/tmp/testing/artifacts/salt/lib/python3.10/site-packages/distro.py", line 556, in __get__
    ret = obj.__dict__[self._fname] = self._f(obj)
  File "/tmp/testing/artifacts/salt/lib/python3.10/site-packages/distro.py", line 1014, in _lsb_release_info
    stdout = subprocess.check_output(cmd, stderr=devnull)
  File "/tmp/testing/artifacts/salt/lib/python3.10/subprocess.py", line 421, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/tmp/testing/artifacts/salt/lib/python3.10/subprocess.py", line 526, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '('lsb_release', '-a')' returned non-zero exit status 126.
```

Would you be happy with that?

Reply:
:+1: yeah, fix / work-around should be in linux_distribution not in callers to it.